rpc/src/helpers.rs

```diff
@@ -527,4 +859,88 @@ pub fn get_pseudo_random_number(state: RandomSeedState, bound: i32) -> TezosPRNG
             };
         }
         Ok((v.into(), sequence))
+}
+
+/// Return block level based on block_id url parameter
+/// 
+/// # Arguments
+/// 
+/// * `block_id` - Url parameter block_id.
+/// * `persistent_storage` - Persistent storage handler.
+/// * `state` - Current RPC collected state (head).
+/// 
+/// If block_id is head return current head level
+/// If block_id is level then return level as i64
+/// if block_id is block hash string return level from BlockMetaStorage by block hash string
+#[inline]
+pub(crate) fn get_level_by_block_id(block_id: &str, persistent_storage: &PersistentStorage, state: &RpcCollectedStateRef) -> Result<Option<usize>, failure::Error> {
+    // first try to parse level as number
+    let level = match block_id.parse() {
+        // block level was passed as parameter to block_id
+        Ok(val) => Some(val),
+        // block hash string or 'head' was passed as parameter to block_id
+        Err(_e) => {
+            let block_hash = get_block_hash_by_block_id(block_id, persistent_storage, state)?;
+            let block_meta_storage: BlockMetaStorage = BlockMetaStorage::new(persistent_storage);
+            if let Some(block_meta) = block_meta_storage.get(&block_hash)? {
+                Some(block_meta.level() as usize)
+            } else {
+                None
+            }
+        }
+    };
+
+    Ok(level)
+}
+
+/// Get block has bytes from block hash or block level
+/// # Arguments
+/// 
+/// * `block_id` - Url parameter block_id.
+/// * `persistent_storage` - Persistent storage handler.
+/// * `state` - Current RPC collected state (head).
+/// 
+/// If block_id is head return block hash byte string from current RpcCollectedStateRef
+/// If block_id is level then return block hash byte string from BlockStorage by level
+/// if block_id is block hash string return block hash byte string from BlockStorage by block hash string
+#[inline]
+pub(crate) fn get_block_hash_by_block_id(block_id: &str, persistent_storage: &PersistentStorage, state: &RpcCollectedStateRef) -> Result<BlockHash, failure::Error> {
```

Comment:
I have unified to pass RpcCollectedStateRef as reference to all functions instead of cloning it sometimes. Is it OK or should I rework it back to not pass reference and clone it if necessary?

Reply:
Good job! Pass by reference by default and clone if necessary is a good strategy here.